src/v/kafka/server/handlers/describe_configs.cc

```diff
@@ -77,12 +79,80 @@ static ss::sstring describe_as_string(const T& t) {
     return ssx::sformat("{}", t);
 }
 
+// property_config_type maps the datatype for a config property to
+// describe_configs_type. Currently class_type and password are not used in
+// Redpanda so we do not include checks for those types. You may find a similar
+// mapping in Apache Kafka at
+// https://github.com/apache/kafka/blob/be032735b39360df1a6de1a7feea8b4336e5bcc0/core/src/main/scala/kafka/server/ConfigHelper.scala
+// Type sizes are based on Java and taken from
+// https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
+template<typename T>
+describe_configs_type property_config_type(std::string_view name) {
+    // Some properties, such as delete_retention_ms, are converted to a string
+    // in one of the overloaded add_topic_configs. So specifically check
+    // for those property names.
+    if (name == config::shard_local_cfg().delete_retention_ms.name()) {
+        return describe_configs_type::long_type;
+    } else if (name == config::shard_local_cfg().retention_bytes.name()) {
+        return describe_configs_type::long_type;
+    } else if (name == topic_property_retention_local_target_bytes) {
+        return describe_configs_type::long_type;
+    } else if (name == topic_property_retention_local_target_ms) {
+        return describe_configs_type::long_type;
+    } else if constexpr (reflection::is_std_optional<T>) {
+        return property_config_type<typename T::value_type>(name);
+    } else if constexpr (std::is_same_v<T, bool>) {
+        return describe_configs_type::boolean;
+    } else if constexpr (std::is_same_v<T, ss::sstring>) {
+        return describe_configs_type::string;
```

Comment:
i find this to be confusing. what are the semantics of mixing constexpr and non-constexpr conditionals? do they all then become non-constexpr?

i could also do something like `property_config_type<std::vector<std::string>>(topic_property_retention_local_target_bytes)` and it looks like i'd get back `long_type` which is confusing.

perhaps this should be two separate functions? one for names and one for types? maybe i'm missing some context?